### ISSUE TITLE
An easier approach to building table components or am I missing something?

### ISSUE THREAD

**Issue by thephoenixofthevoid** [tannerlinsley/react-table#1865]
Consider **material-UI-components** example (I am skipping all imports and not relevant parts):
```ts
function Table({ columns, data }) {
  // Use the state and functions returned from useTable to build your UI
  const { getTableProps, headerGroups, rows, prepareRow } = useTable({
    columns,
    data,
  })

  // Render the UI for your table
  return (
    <MaUTable {...getTableProps()}>
      <TableHead>
        {headerGroups.map(headerGroup => (
          <TableRow {...headerGroup.getHeaderGroupProps()}>
            {headerGroup.headers.map(column => (
              <TableCell {...column.getHeaderProps()}>
                {column.render('Header')}
              </TableCell>
            ))}
          </TableRow>
        ))}
      </TableHead>
      <TableBody>
        {rows.map((row, i) => {
          prepareRow(row)
          return (
            <TableRow {...row.getRowProps()}>
              {row.cells.map(cell => {
                return (
                  <TableCell {...cell.getCellProps()}>
                    {cell.render('Cell')}
                  </TableCell>
                )
              })}
            </TableRow>
          )
        })}
      </TableBody>
    </MaUTable>
  )
}
```
It's rather a complicated piece of code, hard to reason about and, on top of that, it's using inline functions that will be recreated each rerender.

I ended with an alternative approach:
```ts
const h = React.createElement;

function renderCell<T extends object>(cell: Cell<T>) {
  return h(TableCell, cell.getCellProps(), cell.render('Cell'))
}

function renderHeader<T extends object>(column: ColumnInstance<T>) {
  return h(TableCell, column.getHeaderProps(), column.render('Header'))
}

function renderHeaderGroup<T extends object>(g: HeaderGroup<T>) {
  return h(TableRow, g.getHeaderGroupProps(), g.headers.map(renderHeader))
}

function renderRow<T extends object>(this: TableInstance<T>, row: Row<T>, i?: number, rows?: Row<T>[]) {
  this.prepareRow(row);
  return h(TableRow, row.getRowProps(), row.cells.map(renderCell))
}

function Table<T extends object>(props) {
  const instance = useTable<T>(props);
  
  return h(MaUTable, instance.getTableProps(), [
    <TableHead children={instance.headerGroups.map(renderHeaderGroup)}/>,
    <TableBody children={instance.rows.map(renderRow.bind(instance))}/>
  ])
}
```
The logic of functions doesn't change comparing to respective arrow functions from the original one, but without any special treatment they stay referentially the same (except for binding instance to renderRow).

You may notice that I am mixing using `createElement` directly and JSX. This is because using destructuring of an object resulting from a method invocation inside JSX looks cumbersome. Just passing it to the second argument of `h` looks not.

I am wondering if I am wrong at some point and cannot see where things go wrong? And if not, why this approach (seeming natural to me) has not been mentioned anywhere yet?

Thx

**Comment by tannerlinsley** [comment]
I’ll post this here: https://kentcdodds.com/blog/answers-to-common-questions-about-render-props. It’s not all the way up to date (doesn’t talk about hooks) but the concepts are the same.

TLDR, JavaScript is way faster than that. An in-line function is not a big deal at all. Extracting them out for performance is nothing more than an unmeasured preoptimization and is widely considered unnecessary.

**Comment by thephoenixofthevoid** [comment]
It's not only about performance, but to much extend about cognitive overhead when reasoning about 4-5 level nested JSX. Considering performance, the difference is subtle, but can be perceived, when large table is loaded without virtualisation. But that's not my primary point. 

Let me illustrate my primary point by getting rid of ts annotations:
```js
const h = React.createElement;

function renderCell(cell) {
  return h(TableCell, cell.getCellProps(), cell.render('Cell'))
}

function renderHeader(column) {
  return h(TableCell, column.getHeaderProps(), column.render('Header'))
}

function renderHeaderGroup(group) {
  return h(TableRow, group.getHeaderGroupProps(), group.headers.map(renderHeader))
}

function renderRow(row) {
  this.prepareRow(row);
  return h(TableRow, row.getRowProps(), row.cells.map(renderCell))
}

function Table(props) {
  const instance = useTable(props);
  return h(MaUTable, instance.getTableProps(), [
    <TableHead children={instance.headerGroups.map(renderHeaderGroup)}/>,
    <TableBody children={instance.rows.map(renderRow.bind(instance))}/>
  ])
}
```

**Comment by tannerlinsley** [comment]
No worries on removing the TS. First I would ask that you provide stats and reproducible measurements to demonstrate that the difference is non negligible. People write nested JSX and in-line map functions all day with no issues. I’ve never experienced measurable performance issues related to this pattern, even in datavis scenarios rendering thousands of elements.

Now, if you prefer to extract just for the sake of mental overhead/style/preference, then go for it! Again I don’t see an issue with the nesting of JSX. Contrary to how I prefer to write all of my normal code (extracted and functional), I prefer it when writing markup and JSX.

**Comment by tannerlinsley** [comment]
Feel free to discuss more, but going to close this for now.
